### PR TITLE
test

### PR DESCRIPTION
**Pull request by levnach**
It is just a test. Do not merge

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -96,7 +96,7 @@ add_subdirectory(tactic/portfolio)
 add_subdirectory(opt)
 add_subdirectory(api)
 add_subdirectory(api/dll)
-
+add_definitions(-DDUMP_ARGS)
```
Comment (NikolajBjorner) [inline src/CMakeLists.txt]: what is use case for dump-args? make it a global configuration option instead?

File: src/nlsat/nlsat_explain.cpp [hunk 2]
```diff
@@ -1876,7 +1876,7 @@ void pp(nlsat::explain::imp & ex, polynomial_ref_vector const & ps) {
     ex.display(std::cout, ps);
 }
 void pp_var(nlsat::explain::imp & ex, nlsat::var x) {
-    ex.display(std::cout, x);
+    //    ex.display(std::cout, x);
```
Comment (NikolajBjorner) [inline src/nlsat/nlsat_explain.cpp]: delete line

File: src/shell/main.cpp [hunk 3]
```diff
@@ -299,7 +299,7 @@ static void parse_cmd_line_args(int argc, char ** argv) {
 
 
 int STD_CALL main(int argc, char ** argv) {
-#ifdef DUMP_ARGS
+#ifdef DUMP_ARGS_
```
Comment (NikolajBjorner) [inline src/shell/main.cpp]: I would delete this too

File: src/tactic/smtlogics/CMakeLists.txt [hunk 4]
```diff
@@ -40,3 +40,4 @@ z3_add_component(smtlogic_tactics
     qfufbv_tactic.h
     quant_tactics.h
 )
+add_definitions(-DSTART_ONLY_QFNIA_SMT)
```
Comment (NikolajBjorner) [inline src/tactic/smtlogics/CMakeLists.txt]: remove

File: src/test/lp/lp.cpp [hunk 5]
```diff
@@ -2826,8 +2836,8 @@ void test_bound_propagation_one_row_with_bounded_vars() {
     unsigned x0 = ls.add_var(0, false);
     unsigned x1 = ls.add_var(1, false);
     vector<std::pair<mpq, var_index>> c;
-    c.push_back(std::pair<mpq, var_index>(1, x0));
-    c.push_back(std::pair<mpq, var_index>(-1, x1));
+    c.push_back(std::pair<mpq, var_index>(mpq(1), x0));
+    c.push_back(std::pair<mpq, var_index>(-mpq(1), x1));
```
Comment (NikolajBjorner) [inline src/test/lp/lp.cpp]: mpq(-1)?

File: src/util/lp/factorization.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    init_vars_by_the_mask(k_vars, j_vars);
+    SASSERT(!k_vars.empty() && !j_vars.empty());
+    std::sort(k_vars.begin(), k_vars.end());
+    std::sort(j_vars.begin(), j_vars.end());
+
+    if (k_vars.size() == 1) {
+        k.index() = k_vars[0];
+        k.type() = factor_type::VAR;
+    } else {
+        unsigned i;
```
Comment (NikolajBjorner) [inline src/util/lp/factorization.cpp]: i = 0 to avoid compiler warnings

File: src/util/lp/factorization.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        k.index() = i;
+        k.type() = factor_type::RM;
+    }
+
+    if (j_vars.size() == 1) {
+        j.index() = j_vars[0];
+        j.type() = factor_type::VAR;
+    } else {
+        unsigned i;
```
Comment (NikolajBjorner) [inline src/util/lp/factorization.cpp]: i = 0; to avoid compiler warnings

File: src/util/lp/factorization.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  Module Name:
+
+  <name>
+
+  Abstract:
+
+  <abstract>
+
+  Author:
+  Nikolaj Bjorner (nbjorner)
```
Comment (NikolajBjorner) [inline src/util/lp/factorization.h]: did I actually contribute to this file?

File: src/util/lp/factorization.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+class factor {
+    unsigned     m_index;
+    factor_type  m_type;
+public:
+    factor() {}
+    explicit factor(unsigned j) : factor(j, factor_type::VAR) {}
+    factor(unsigned i, factor_type t) : m_index(i), m_type(t) {}
+    unsigned index() const {return m_index;}
+    unsigned& index() {return m_index;}
```
Comment (NikolajBjorner) [inline src/util/lp/factorization.h]: style: space or not to space? Use space after { is used elsewhere.

File: src/util/lp/factorization.h [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+struct const_iterator_mon {
+    // fields
+    svector<bool>                  m_mask;
+    const factorization_factory *  m_ff;
+    bool                           m_full_factorization_returned;
+
+    //typedefs
+    typedef const_iterator_mon self_type;
+    typedef factorization value_type;
+    typedef const factorization reference;
```
Comment (NikolajBjorner) [inline src/util/lp/factorization.h]: why is this called "reference"?

File: src/util/lp/lar_solver.h [hunk 11]
```diff
@@ -148,8 +159,12 @@ public :
 
     var_index add_var(unsigned ext_j, bool is_integer);
 
+    var_index add_named_var(unsigned ext_j, bool is_integer, std::string);
```
Comment (NikolajBjorner) [inline src/util/lp/lar_solver.h]: pass std::string const& by reference

File: src/util/lp/lar_solver.h [hunk 12]
```diff
@@ -420,34 +470,34 @@ public :
...
 
-
     void get_model(std::unordered_map<var_index, mpq> & variable_values) const;
 
     void get_model_do_not_care_about_diff_vars(std::unordered_map<var_index, mpq> & variable_values) const;
 
     std::string get_variable_name(var_index vi) const;
 
-    // print utilities
+    void set_variable_name(var_index vi, std::string);
```
Comment (NikolajBjorner) [inline src/util/lp/lar_solver.h]: pass by reference ?

File: src/util/lp/lp_core_solver_base_def.h [hunk 13]
```diff
@@ -698,6 +698,9 @@ non_basis_is_correctly_represented_in_heading() const {
 
 template <typename T, typename X> bool lp_core_solver_base<T, X>::
     basis_heading_is_correct() const {
+    if ( m_A.column_count() > 10 ) {
```
Comment (NikolajBjorner) [inline src/util/lp/lp_core_solver_base_def.h]: temporary? for perf?

File: src/util/lp/monomial.h [hunk 14]
```diff
@@ -1,33 +1,66 @@
-/*++
-Copyright (c) 2017 Microsoft Corporation
-
-Module Name:
+/*
+  Copyright (c) 2017 Microsoft Corporation
+  Author: Nikolaj Bjorner
```
Comment (NikolajBjorner) [inline src/util/lp/monomial.h]: No, you cannot hide.

File: src/util/lp/nla_solver.h [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    lp::lar_term         m_term;
+    rational             m_rs;
+    ineq(lp::lconstraint_kind cmp, const lp::lar_term& term, const rational& rs) : m_cmp(cmp), m_term(term), m_rs(rs) {}
+    bool operator==(const ineq& a) const {
+        return m_cmp == a.m_cmp && m_term == a.m_term && m_rs == a.m_rs;
+    }
+    const lp::lar_term& term() const {
+        return m_term;
+    };
+    const lp::lconstraint_kind& cmp() const {
```
Comment (NikolajBjorner) [inline src/util/lp/nla_solver.h]: overkill? just lp::lconstraint_kind

File: src/util/lp/rooted_mons.h [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+    void push_back(const index_with_sign& ind_s) {
+        m_mons.push_back(ind_s);
+    }
+};
+
+
+struct rooted_mon_table {
+    std::unordered_map<svector<lpvar>,
+                       unsigned, // points to vec()
+                       hash_svector>                                 m_map;
```
Comment (NikolajBjorner) [inline src/util/lp/rooted_mons.h]: rename m_map / m_vector to something specific/descriptive?

File: src/util/lp/rooted_mons.h [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+    vector<rooted_mon>                                               m_vector;
+    // for every k 
+    // for each i in m_rooted_monomials_containing_var[k]
+    // m_vector_of_rooted_monomials[i] contains k
+    std::unordered_map<lpvar, std::unordered_set<unsigned>>          m_mons_containing_var;
+
+    // A map from m_vector_of_rooted_monomials to a set
```
Comment (NikolajBjorner) [inline src/util/lp/rooted_mons.h]: set of sets or just set?

File: src/util/lp/rooted_mons.h [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+        double r = s /m_map.size();
+        if (r > ratio) {
+            ratio = r;
+            out << "rooted mons "  << m_map.size() << ", ratio = " << r << "\n";
+        }
+    }
+
+    const svector<unsigned>& to_refine() { return m_to_refine; }
+
+    bool list_is_consistent(const vector<index_with_sign>& list, const std::unordered_set<unsigned>& to_refine_reg) const {
```
Comment (NikolajBjorner) [inline src/util/lp/rooted_mons.h]: work with empty list?
for (unsigned i = 0, j = 0; i < list.size() && (j == 0 || i == j); ++i) {
      if (to_refine_reg.contains(list[i].index())) j++;
}
return j == 0 || j == list.size();

File: src/util/lp/rooted_mons.h [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+    std::unordered_map<unsigned, std::unordered_set<unsigned>>& proper_factors() {
+        return m_proper_factors;
+    }
+
+    const std::unordered_map<unsigned, std::unordered_set<unsigned>>& proper_factors() const {
+        return m_proper_factors;
+    }
+
+    void reduce_set_by_checking_proper_containment(std::unordered_set<unsigned>& p,
```
Comment (NikolajBjorner) [inline src/util/lp/rooted_mons.h]: elsewhere, we tend to use pattern:

auto j = p.begin();
for (auto i = p.begin(); i != p.end(); ++i) 
   if (test(i)) *j++ =*i;
p.set_end(j);

File: src/util/lp/rooted_mons.h [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+                continue;
+            }
+            auto iit = it;
+            iit ++;
+            p.erase(it);
+            it = iit;
+        }
+    }
+
+    // here i is the index of a monomial in m_vector
```
Comment (NikolajBjorner) [inline src/util/lp/rooted_mons.h]: assertion that rm is non-empty?

File: src/util/lp/vars_equivalence.h [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+        auto it = m_tree.find(j);
+        if (it == m_tree.end())
+            return r;
+
+        const node& n = it->second;
+        for (unsigned e_k: n.m_children) {
+            const equiv & e = m_equivs[e_k];
+            lpvar oj = e.m_i == j? e.m_j : e.m_i;
+            r.push_back(oj);
+            for (lpvar k :  children(oj)) {
```
Comment (NikolajBjorner) [inline src/util/lp/vars_equivalence.h]: is this shallow/non-shared?

File: src/util/lp/vars_equivalence.h [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+
+    struct node {
+        unsigned            m_parent;   // points to m_equivs
+        svector<unsigned>   m_children;  // point to m_equivs
+        node() : m_parent(-1) {}
+    };
+   
+    //fields
+    
+    // The map from the variables to m_nodes
```
Comment (NikolajBjorner) [inline src/util/lp/vars_equivalence.h]: use union-find data-structures with embedded doubly linked lists for equivalence class traversal?

File: src/util/lp/vars_equivalence.h [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+typedef lp::var_index            lpvar;
+struct hash_svector {
+    size_t operator()(const unsigned_vector & v) const {
+        return svector_hash<unsigned_hash>()(v);
+    }
+};
+
+
+
+struct vars_equivalence {
```
Comment (NikolajBjorner) [inline src/util/lp/vars_equivalence.h]: are all methods required to be public?

File: src/util/lp/vars_equivalence.h [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+        svector<unsigned>   m_children;  // point to m_equivs
+        node() : m_parent(-1) {}
+    };
+   
+    //fields
+    
+    // The map from the variables to m_nodes
+    // m_tree is a spanning tree of the graph of equivs represented by m_equivs
+    std::unordered_map<unsigned, node>            m_tree;  
+    // If m_tree[v] == -1 then the variable is a root.
```
Comment (NikolajBjorner) [inline src/util/lp/vars_equivalence.h]: would be useful with a more abstract spec of what this class achieves.
The many lookups to m_tree seem internal, not tied to the spec, and could be replaced with a 1-1 relation of nodes and indices in m_equivs or related.

File: src/util/rational.h [hunk 25]
```diff
@@ -53,7 +53,7 @@ class rational {
 
     rational(mpz const & z) { m().set(m_val, z); }
 
-    rational(double  z) { UNREACHABLE(); }
+    explicit rational(double  z) { UNREACHABLE(); }
```
Comment (NikolajBjorner) [inline src/util/rational.h]: yes, this is good. Can you push this one to master right away?

File: src/util/lp/factorization.cpp (hunk 7 above)
Comment (levnach) [reply src/util/lp/factorization.cpp]: I am not seeing a warning here

File: src/util/lp/factorization.h (hunk 8 above)
Comment (levnach) [reply src/util/lp/factorization.h]: I believe there were some discussions about it with you.

File: src/util/lp/factorization.h (hunk 10 above)
Comment (levnach) [reply src/util/lp/factorization.h]: Removed

File: src/util/lp/lp_core_solver_base_def.h (hunk 13 above)
Comment (levnach) [reply src/util/lp/lp_core_solver_base_def.h]: for performance

File: src/util/lp/rooted_mons.h (hunk 17 above)
Comment (levnach) [reply src/util/lp/rooted_mons.h]: I added a comprehensive comment.

File: src/util/lp/vars_equivalence.h (hunk 21 above)
Comment (levnach) [reply src/util/lp/vars_equivalence.h]: Sorry, I do not understand the question.

File: src/util/lp/vars_equivalence.h (hunk 22 above)
Comment (levnach) [reply src/util/lp/vars_equivalence.h]: Postpone the revision of this class

File: src/util/lp/vars_equivalence.h (hunk 23 above)
Comment (levnach) [reply src/util/lp/vars_equivalence.h]: Postpone the revision of this class

File: src/util/lp/vars_equivalence.h (hunk 24 above)
Comment (levnach) [reply src/util/lp/vars_equivalence.h]: Postpone the revision of this class

**Pull request closed without merging**